### PR TITLE
fix missing gc-roots

### PR DESCRIPTION
**Pull request by vtjnash**
since calling `pointer()` explicitly discards gc-rooting
need to ensure that there is a memory barrier that uses the value
after we know for certain we are done with the value

### CODE REVIEW

File: src/JLD.jl [hunk 1]
```diff
@@ -208,10 +209,8 @@ function jldopen(filename::AbstractString, rd::Bool, wr::Bool, cr::Bool, tr::Boo
             end
             if startswith(magic, Vector{UInt8}(magic_base))
                 version = convert(VersionNumber, unsafe_string(pointer(magic) + length(magic_base)))
+                gcuse(magic)
                 if version < v"0.1.0"
-                    if !isdefined(JLD, :JLD00)
-                        eval(:(include(joinpath($(dirname(@__FILE__)), "JLD00.jl"))))
```
Comment (tkelman) [inline src/JLD.jl]: must be some reason this had been structured this way? @simonster @timholy either of you remember?

Comment (simonster) [inline src/JLD.jl]: I think this predates precompilation, and the reason was package load time.

Comment (tkelman) [inline src/JLD.jl]: okay sounds obsolete then, thanks

**Comment by JeffBezanson** [comment]
Bump!

**Comment by simonster** [comment]
I'd feel more comfortable merging this if the JLD tests passed on 0.6. (The JLDArchives tests have been broken forever, so I guess no one cares about those, but at least the package tests should pass...)

**Comment by ararslan** [comment]
> The JLDArchives tests have been broken forever, so I guess no one cares about those

If they keep failing and no one cares to fix them, should we just disable them here?

**Comment by simonster** [comment]
No. We should fix the tests. While there's an argument to be made that we should drop support for really old JLD files, there should be some kind of testing to make sure that JLD can actually read files created with previous commits.

**Comment by vtjnash** [comment]
This fixes all of the tests, but only on master. Looks like something is different enough about v0.6 that it needed something else though.

File: src/JLD.jl [hunk 2]
```diff
@@ -870,13 +875,13 @@ JLD.writeas(data::StackFrame) =
 const _where_macrocall = Symbol("@where")
 function expand_where_macro(e::Expr)
     e.head = :where
-    shift!(e.args)
```
Comment (ararslan) [inline src/JLD.jl]: Did I mess this up in the rebase? That is, is this line supposed to remain and the args are shifted again if macros have sourceloc?

File: src/JLD.jl [hunk 3]
```diff
@@ -953,13 +990,7 @@ end
 function julia_type(e::Union{Symbol, Expr})
     if is_valid_type_ex(e)
         try # `try` needed to catch undefined symbols
-            typ = eval(current_module(), e)
```
Comment (simonster) [inline src/JLD.jl]: Removing this is what broke the test, but I fully agree that no one should be relying on this behavior.

**Comment by simonster** [comment]
The tests now pass on 0.6 but not 0.7. It looks like we will eventually have to eliminate the `mmaparrays` option, since there's no guarantee that arbitrary arrays in a JLD file will meet the new alignment requirement.

**Pull request merged**